src/message/dkim.rs

```diff
@@ -0,0 +1,348 @@
+        }
+    }
+}
+
+/// Describe Canonicalization to be applied before signing
+#[derive(Copy, Clone)]
+pub struct DkimCanonicalization {
+    header: DkimCanonicalizationType,
+    body: DkimCanonicalizationType,
+}
+
+/// Format canonicalization to be shown in Dkim header
+impl Display for DkimCanonicalization {
+    fn fmt(&self, fmt: &mut std::fmt::Formatter<'_>) -> std::result::Result<(), std::fmt::Error> {
+        write!(fmt, "{}/{}", self.header, self.body)
+    }
+}
+
+/// A struct to describe Dkim configuration applied when signing a message
+/// selector: the name of the key publied in DNS
+/// domain: the domain for which we sign the message
+/// private_key: private key in PKCS1 string format
+/// headers: a list of headers name to be included in the signature. Signing of more than one
+/// header with same name is not supported
+/// canonicalization: the canonicalization to be applied on the message
+#[derive(Clone)]
+pub struct DkimConfig {
+    selector: String,
+    domain: String,
+    private_key: String,
+    headers: Vec<String>,
+    canonicalization: DkimCanonicalization,
+}
+
+impl DkimConfig {
+    /// Create a default signature configuration with a set of headers and "simple/relaxed"
+    /// canonicalization
+    pub fn default_config(selector: String, domain: String, private_key: String) -> DkimConfig {
+        DkimConfig {
+            selector,
+            domain,
+            private_key,
+            headers: vec![
+                "From".to_string(),
+                "Subject".to_string(),
+                "To".to_string(),
+                "Date".to_string(),
+            ],
+            canonicalization: DkimCanonicalization {
+                header: DkimCanonicalizationType::Simple,
+                body: DkimCanonicalizationType::Relaxed,
+            },
+        }
+    }
+}
+
+/// Create a Headers struct with a Dkim-Signature Header created from given parameters
+fn dkim_header_format(
+    domain: String,
+    selector: String,
+    canon: DkimCanonicalization,
+    timestamp: String,
+    headers_list: String,
+    body_hash: String,
+    signature: String,
+) -> Headers {
+    let mut headers = Headers::new();
+    let header_name = match canon.header {
+        DkimCanonicalizationType::Simple => HeaderName::new_from_ascii_str("DKIM-Signature"),
+        DkimCanonicalizationType::Relaxed => HeaderName::new_from_ascii_str("dkim-signature"),
+    };
+    headers.append_raw(header_name, format!("v=1; a=rsa-sha256; d={domain}; s={selector}; c={canon}; q=dns/txt; t={timestamp}; h={headers_list}; bh={body_hash}; b={signature}",domain=domain, selector=selector,canon=canon,timestamp=timestamp,headers_list=headers_list,body_hash=body_hash,signature=signature));
+    headers
+}
+
+/// Canonicalize the body of an email
+fn dkim_canonicalize_body(body: &[u8], canonicalization: DkimCanonicalizationType) -> String {
+    let body = std::str::from_utf8(body).unwrap();
```

Comment:
The email body almost always is utf-8 but it's totally possible to build a non utf-8 body https://docs.rs/lettre/0.10.0-rc.3/lettre/message/struct.Body.html#method.dangerous_pre_encoded

Reply:
I use this function withthe body as displayed by formatted method for the message, is this possible to be non utf-8 ?
Is there a method to get body as before being sent ?